### PR TITLE
Full file support for roughness ini FrictFile

### LINKED ISSUE #118: Support roughness.ini files
https://content.oss.deltares.nl/delft3d/manuals/D-Flow_FM_User_Manual_1D2D.pdf#section.C.15

### PR DESCRIPTION
**Pull request by arthurvd**
Fixes #118

### CODE REVIEW

File: hydrolib/core/io/friction/models.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            "numLevels lines containing space separated lists of roughness values: "
+            + "numLocations values per line. If the functionType is constant, then a "
+            + "single line is required. For a uniform roughness per branch "
+            + "(numLocations = 0) a single entry per line is required. The meaning "
+            + "of the values depends on the roughness type selected (see frictionType).",
+            alias="frictionValues",
+        )
+
+    comments: Comments = Comments()
+    _header: Literal["Branch"] = "Branch"
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:107]: Isnt the header name "BranchProperties"?

Comment (arthurvd) [reply hydrolib/core/io/friction/models.py:107]: No, it's "Branch", see table C.19 in https://content.oss.deltares.nl/delft3d/manuals/D-Flow_FM_User_Manual_1D2D.pdf#section.C.15

File: hydrolib/core/io/friction/models.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+from pathlib import Path
+from typing import List, Literal, Optional, Set, Union
+
+from pydantic import Field
+from pydantic.class_validators import root_validator, validator
+
+from hydrolib.core.io.ini.models import INIBasedModel, INIGeneral, INIModel
+from hydrolib.core.io.ini.util import (
+    get_enum_validator,
+    get_from_subclass_defaults,
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:11]:
```suggestion
```

File: hydrolib/core/io/friction/models.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    get_split_string_on_delimiter_validator,
+)
+from hydrolib.core.utils import str_is_empty_or_none
+
+from ..ini.util import make_list_validator
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:16]:
```suggestion
import logging
from pathlib import Path
from typing import List, Literal, Optional

from pydantic import Field

from hydrolib.core.io.ini.models import INIBasedModel, INIGeneral, INIModel
from hydrolib.core.io.ini.util import get_split_string_on_delimiter_validator

from ..ini.util import make_list_validator
```

File: hydrolib/core/io/friction/models.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    frictiontype: str = Field(alias="frictionType")
+    functiontype: Optional[str] = Field("constant", alias="functionType")
+    timeseriesid: Optional[str] = Field(alias="timeSeriesId")
+    numlevels: Optional[int] = Field(alias="numLevels")
+    levels: Optional[List[float]]
+    numlocations: Optional[int] = Field(0, alias="numLocations")
+    chainage: Optional[List[float]]
+    frictionvalues: Optional[List[float]] = Field(
+        alias="frictionValues"
+    )  # TODO: turn this into List[List[float]]
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:118]: Do you still want to solve this TODO?

File: hydrolib/core/io/friction/models.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    get_from_subclass_defaults,
+    get_split_string_on_delimiter_validator,
+)
+from hydrolib.core.utils import str_is_empty_or_none
+
+from ..ini.util import make_list_validator
+
+logger = logging.getLogger(__name__)
+
+
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:20]: In the remarks section of [https://content.oss.deltares.nl/delft3d/manuals/D-Flow_FM_User_Manual_1D2D.pdf#section.C.15](https://content.oss.deltares.nl/delft3d/manuals/D-Flow_FM_User_Manual_1D2D.pdf#section.C.15) there are some rules, do we want to add validation for this, perhaps in another issue?

File: hydrolib/core/io/friction/models.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    get_enum_validator,
+    get_from_subclass_defaults,
+    get_split_string_on_delimiter_validator,
+)
+from hydrolib.core.utils import str_is_empty_or_none
+
+from ..ini.util import make_list_validator
+
+logger = logging.getLogger(__name__)
+
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:19]: I think we need some simple tests for the friction classes, to make sure the instances can be properly initialized with kwargs, +/- optional parameters.

**Review by priscavdsluis — CHANGES_REQUESTED** [review]
Some comments @arthurvd

File: hydrolib/core/io/friction/models.py (hunk 4 above)
Comment (arthurvd) [reply hydrolib/core/io/friction/models.py:112]: After our discussion this afternoon, this is now under a new issue #143.

File: hydrolib/core/io/friction/models.py (hunk 6 above)
Comment (arthurvd) [reply hydrolib/core/io/friction/models.py:14]: See 98a1e47.

**Comment by arthurvd** [comment]
> Some comments @arthurvd

@priscavdsluis : please re-review my latest commits for a bugfix, and several validation + test functions.

File: hydrolib/core/io/friction/models.py [hunk 7]
```diff
@@ -109,13 +104,13 @@ class Comments(INIBasedModel.Comments):
     frictiontype: str = Field(alias="frictionType")
     functiontype: Optional[str] = Field("constant", alias="functionType")
     timeseriesid: Optional[str] = Field(alias="timeSeriesId")
-    numlevels: Optional[int] = Field(alias="numLevels")
+    numlevels: Optional[PositiveInt] = Field(alias="numLevels")
     levels: Optional[List[float]]
-    numlocations: Optional[int] = Field(0, alias="numLocations")
+    numlocations: Optional[NonNegativeInt] = Field(0, alias="numLocations")
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:109]: Nice!

File: hydrolib/core/io/friction/models.py [hunk 8]
```diff
@@ -124,15 +119,59 @@ class Comments(INIBasedModel.Comments):
...
 class FrictionModel(INIModel):
     general: FrictGeneral = FrictGeneral()
-    global_: FrictGlobal = Field(alias="global")  # to circumvent built-in kw
+    global_: List[FrictGlobal] = Field([], alias="global")  # to circumvent built-in kw
     branch: List[FrictBranch] = []
 
     _split_to_list = make_list_validator(
+        "global_",
+    )
+    _split_to_list2 = make_list_validator(
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:174]: Why was this changed?

File: hydrolib/core/io/friction/models.py [hunk 9]
```diff
@@ -124,15 +119,59 @@ class Comments(INIBasedModel.Comments):
...
+        # number of values should be equal to numlocations*numlevels
+        numvals = max(1, values["numlocations"]) * (
+            1
+            if (
+                "numlevels" not in values
+                or values["numlevels"] is None
+                or values["numlevels"] is 0
+            )
+            else values["numlevels"]
+        )
```
Comment (priscavdsluis) [inline hydrolib/core/io/friction/models.py:157]:
```suggestion
        numlevels = (
            1
            if (
                "numlevels" not in values
                or values["numlevels"] is None
                or values["numlevels"] is 0
            )
            else values["numlevels"]
        )
        numvals = max(1, values["numlocations"]) * numlevels
```

File: hydrolib/core/io/friction/models.py (hunk 8 above)
Comment (arthurvd) [reply hydrolib/core/io/friction/models.py:174]: The file format in https://content.oss.deltares.nl/delft3d/manuals/D-Flow_FM_User_Manual_1D2D.pdf#section.C.15 defines:
> Multiple roughness variable definitions that don’t vary spatially (i.e. that don’t require Branch blocks) may be combined in one file. In that case there will one General block, multiple instances of the Global block and no Branch blocks in the file. See the second example.

So, our ```FrictionModel.global_``` must be a list, and not a scalar.

Comment (priscavdsluis) [reply hydrolib/core/io/friction/models.py:175]: Sorry, I was unclear, what I meant was: you don't need to call make_list_validator twice you can just do the following:

```
    _split_to_list = make_list_validator(
        "global_",
        "branch",
    )
```

**Review by priscavdsluis — APPROVED** [review]
